src/nvim/os/pty_process.c

```diff
@@ -0,0 +1,225 @@
+// Some of the code came from pangoterm and libuv
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <unistd.h>
+#include <termios.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <sys/ioctl.h>
+
+// forkpty is not in POSIX, so headers are platform-specific
+#if defined(__FreeBSD__)
+# include <libutil.h>
+#elif defined(__OpenBSD__) || defined(__NetBSD__) || defined(__APPLE__)
+# include <util.h>
+#else
+# include <pty.h>
+#endif
+
+#include <uv.h>
+
+#include "nvim/os/job.h"
+#include "nvim/os/job_defs.h"
+#include "nvim/os/job_private.h"
+#include "nvim/os/pty_process.h"
+#include "nvim/memory.h"
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "os/pty_process.c.generated.h"
+#endif
+
+typedef struct {
+  struct winsize winsize;
+  uv_pipe_t proc_stdin, proc_stdout, proc_stderr;
+  uv_signal_t schld;
+  int tty_fd;
+} PtyProcess;
+
+void pty_process_init(Job *job)
+{
+  PtyProcess *ptyproc = xmalloc(sizeof(PtyProcess));
+
+  if (job->opts.writable) {
+    uv_pipe_init(uv_default_loop(), &ptyproc->proc_stdin, 0);
+    ptyproc->proc_stdin.data = NULL;
+  }
+
+  if (job->opts.stdout_cb) {
+    uv_pipe_init(uv_default_loop(), &ptyproc->proc_stdout, 0);
+    ptyproc->proc_stdout.data = NULL;
+  }
+
+  if (job->opts.stderr_cb) {
+    uv_pipe_init(uv_default_loop(), &ptyproc->proc_stderr, 0);
+    ptyproc->proc_stderr.data = NULL;
+  }
+
+  job->proc_stdin = (uv_stream_t *)&ptyproc->proc_stdin;
+  job->proc_stdout = (uv_stream_t *)&ptyproc->proc_stdout;
+  job->proc_stderr = (uv_stream_t *)&ptyproc->proc_stderr;
+  job->process = ptyproc;
+}
+
+void pty_process_destroy(Job *job)
+{
+  free(job->opts.term_name);
+  free(job->process);
+  job->process = NULL;
+}
+
+bool pty_process_spawn(Job *job)
+{
+  int master;
+  PtyProcess *ptyproc = job->process;
+  ptyproc->winsize = (struct winsize){job->opts.height, job->opts.width, 0, 0};
+  struct termios termios;
+  init_termios(&termios);
+  uv_disable_stdio_inheritance();
+
+  int pid = forkpty(&master, NULL, &termios, &ptyproc->winsize);
+
+  if (pid < 0) {
+    return false;
+  } else if (pid == 0) {
+    init_child(job);
+    abort();
+  }
+
+  // make sure the master file descriptor is non blocking
+  fcntl(master, F_SETFL, fcntl(master, F_GETFL) | O_NONBLOCK);
+
+  if (job->opts.writable) {
+    uv_pipe_open(&ptyproc->proc_stdin, dup(master));
+  }
+
+  if (job->opts.stdout_cb) {
+    uv_pipe_open(&ptyproc->proc_stdout, dup(master));
+  }
+
+  if (job->opts.stderr_cb) {
+    uv_pipe_open(&ptyproc->proc_stderr, dup(master));
```

Comment:
They are closed by libuv when the stream is closed: https://github.com/libuv/libuv/blob/v1.x/src/unix/stream.c#L1563-L1568